jax/numpy/lax_numpy.py

```diff
@@ -873,6 +873,49 @@ def diff(a, n=1, axis=-1,):
   return a
 
 
+@partial(jit, static_argnums=1)
+def _gradient(a, axis):
+  def gradient_along_axis(a, axis):
+    a_swap = swapaxes(a, 0, axis)
+    a_grad = concatenate((
+      (a_swap[1]  - a_swap[0])[newaxis],
+      (a_swap[2:] - a_swap[:-2]) * 0.5,
+      (a_swap[-1] - a_swap[-2])[newaxis]
+      ), axis=0)
+    return swapaxes(a_grad, 0, axis)
```

Comment:
This is totally fine, but might find it a little cleaner to make use of `jax.lax.slice_in_dim` here instead of using `swapaxes`, i.e.,
```python
    sliced = functools.partial(jax.lax.slice_in_dim, a, axis=axis)
    a_grad = concatenate((
        sliced(1, 2) - sliced(0, 1),
        (sliced(2, None) - sliced(None, -2)) * 0.5,
        sliced(-1, None) - sliced(-2, -1),
    ), axis)
```
I haven't measured this, but it might also result in slightly faster code from XLA.

Reply:
Thank you for that feedback; I was not aware of `jax.lax.slice_in_dim` and this part definitely looks cleaner like this. 

Maybe I am missing something, but `jax.lax.slice_in_dim` neither takes negative indices nor `None`s. Either I plug in that functionality into `jax.lax.slice_in_dim` or this part here gets more messy. I guess the former would not hurt?

Reply:
I've implemented support for negative and `None` indices in `slice_in_dim` and changed `_gradient_along_axis` to your suggestion. 

Performance is indeed better! See this self-contained example: https://gist.github.com/clemisch/f1462d08591eceeb377b31de4eaa2b9e

Reply:
Oh, awesome! Yes, this is my fault for writing code without testing it 😀